### PR TITLE
Fixes #879 - X-Partial Status 

### LINKED ISSUE #879: Correctly set X-Partial-Results
X-Partial-Results and PartialResults in the response should indicate whether we are returning less than the requested page size in a response.

### PR DESCRIPTION
**Pull request by plainolneesh**
Re: #879

### CODE REVIEW

File: web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java [hunk 1]
```diff
@@ -332,11 +331,8 @@ public ResultsPage next() throws Exception {
                 }
             }
         }
-        if (resultList.isEmpty()) {
-            return new ResultsPage();
-        } else {
-            return new ResultsPage(resultList, ((hitPageByteTrigger || hitPageTimeTrigger) ? ResultsPage.Status.PARTIAL : ResultsPage.Status.COMPLETE));
-        }
+        ResultsPage.Status status = (currentPageCount < maxPageSize) ? ResultsPage.Status.PARTIAL : ResultsPage.Status.COMPLETE;
```
Comment (ivakegg) [inline web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java:334]: We are so close here, however a subtlety was missed.  If the page being returned is the last page to be returned, then this is not a "partial page".  So if we have hit the max results, or it.hasNext() is false, then lets return COMPLETE in those cases.

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java [hunk 2]
```diff
@@ -332,13 +331,16 @@ public ResultsPage next() throws Exception {
                 }
             }
         }
-        if (resultList.isEmpty()) {
-            return new ResultsPage();
-        } else {
-            return new ResultsPage(resultList, ((hitPageByteTrigger || hitPageTimeTrigger) ? ResultsPage.Status.PARTIAL : ResultsPage.Status.COMPLETE));
+        ResultsPage.Status status;
+        if (this.maxResults >= numResults) {
```
Comment (ivakegg) [inline web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java:335]: My previous comment was "So if we have hit the max results, or it.hasNext() is false, then lets return COMPLETE in those cases.".  This if statement does not do that.  We "hit the max results" at lines 236 and 241.  In those cases we should return COMPLETE.  Also if !this.it.hasNext()  (i.e it.hasNext() is false) then we should also return COMPLETE.  Only after those checks should you return PARTIAL iff currentPageCount < maxPageSize.  So the code is something like this (provided hitMaxResults is set to true at lines 236 and 241):

if (hitMaxResults || !this.it.hasNext()) {
   status = ResultsPage.Status.COMPLETE;
} else {
   status = (currentPageCount < maxPageSize) ? ResultsPage.Status.PARTIAL : ResultsPage.Status.COMPLETE;
}
return new ResultsPage(resultsList, status);

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java (hunk 2 above)
Comment (ivakegg) [reply web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java:335]: Oh, and BTW this code does not even compile.  Please ensure that all code compiles, passes the formatting, and all tests pass before pushing it up.  Thanks.

Comment (plainolneesh) [reply web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java:335]: I honstly don't know how I made the mistake of pushing this one. The correct one that i wrote was

  if (resultList.isEmpty()) {
            return new ResultsPage();
        } else if (iter.hasNext() && numResults < this.maxResults && currentPageCount < maxPageSize)
            return new ResultsPage(resultList, ResultsPage.Status.PARTIAL);
        else {
            return new ResultsPage(resultList, ResultsPage.Status.COMPLETE);
        }
    }
I am updating this PR now.

File: web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java [hunk 3]
```diff
@@ -26,6 +26,7 @@
 import org.apache.commons.lang.StringEscapeUtils;
 import org.apache.htrace.TraceInfo;
 import org.apache.log4j.Logger;
+import org.infinispan.configuration.parsing.ParseUtils;
```
Comment (lbschanno) [inline web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java:29]: This unused import can be removed.

File: web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java [hunk 4]
```diff
@@ -332,10 +332,13 @@ public ResultsPage next() throws Exception {
                 }
             }
         }
+        
         if (resultList.isEmpty()) {
             return new ResultsPage();
-        } else {
-            return new ResultsPage(resultList, ((hitPageByteTrigger || hitPageTimeTrigger) ? ResultsPage.Status.PARTIAL : ResultsPage.Status.COMPLETE));
+        } else if (iter.hasNext() && numResults < this.maxResults && currentPageCount < maxPageSize)
```
Comment (lbschanno) [inline web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java:338]: The call to `iter.hasNext()` is causing a couple tests in `ExtendedRunningQueryTest` to fail. You'll need to update those tests to expect an additional call to this method. 

```
Error:  Failures: 
Error:    ExtendedRunningQueryTest.testNextMaxResults_HappyPathUsingDeprecatedConstructor:271 
  Unexpected method call TransformIterator.hasNext():
    TransformIterator.hasNext(): expected: 5, actual: 6
Error:    ExtendedRunningQueryTest.testNextWithDnResultLimit_HappyPathUsingDeprecatedConstructor:448 
  Unexpected method call TransformIterator.hasNext():
    TransformIterator.hasNext(): expected: 3, actual: 4
Error:    ExtendedRunningQueryTest.testNext_HappyPathUsingDeprecatedConstructor:184 
  Unexpected method call TransformIterator.hasNext():
    TransformIterator.hasNext(): expected: 3, actual: 4
Error:    ExtendedRunningQueryTest.testNext_NoResultsAfterCancellationUsingDeprecatedConstructor:322 
  Unexpected method call Query.getPagesize():
    Query.getUncaughtExceptionHandler(): expected: 3, actual: 1
    Query.getId(): expected: 2, actual: 1
    Query.getUserDN(): expected: 2, actual: 1
```

**Review by lbschanno — CHANGES_REQUESTED** [review]

File: web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java (hunk 4 above)
Comment (codebyneesh) [reply web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java:337]: Updated the tests please see new commit .

**Review by lbschanno — APPROVED** [review]

**Pull request merged**